### ISSUE TITLE
Number of levels in contour can be larger than the requested number

### ISSUE THREAD

**Issue by mboquien** [matplotlib/matplotlib#9088]
### Bug report

**Bug summary**
The documentation of contour() specifies that contour(X,Y,Z,N) makes a contour with “up to N automatically-chosen levels”. In practice the number of contours can be larger than requested.

**Code for reproduction**

```python
r = linspace(-3., 3., 1000)
x, y = np.meshgrid(r, r)
z = (1. - x/2. + x**5. + y**3.) * np.exp(-x**2. - y**2.)
for i in range(1, 11):
    print("Up to {} levels requested. {} levels provided.".format(i, len(plt.contour(x, y, z, i).levels)))
```

**Actual outcome**

```
Up to 1 levels requested. 2 levels provided.
Up to 2 levels requested. 3 levels provided.
Up to 3 levels requested. 4 levels provided.
Up to 4 levels requested. 4 levels provided.
Up to 5 levels requested. 6 levels provided.
Up to 6 levels requested. 7 levels provided.
Up to 7 levels requested. 7 levels provided.
Up to 8 levels requested. 9 levels provided.
Up to 9 levels requested. 9 levels provided.
Up to 10 levels requested. 9 levels provided.
```

**Expected outcome**

The number of levels provided should never be larger than the number of levels requesting, if the documentation is to be believed. The exact number would be ideal

**Matplotlib version**
  * Operating System: Arch linux
  * Matplotlib Version: 2.0.2
  * Python Version: 3.6.2
Official arch packages.

Note: probably related to issue #4537.

**Comment by efiring** [comment]
I think the solution is to change the "N" in the documentation to "N+1".  The "N" is used as a parameter in MaxNLocator, where it refers to the number of intervals being spanned (i.e., the length), not the number of boundaries.  

The reason MaxNLocator is approximate is that there is a tension among finding "nice" boundaries, not overshooting the data range too much, and being able to specify the number of intervals or boundaries.

If you want to be able to specify N exactly, you are free to supply the levels directly, or to supply a Locator instance that behaves as you desire.

**Comment by jklymak** [comment]
... but couldn't we just pass N-1 to `MaxNLocator`?

**Comment by WeatherGod** [comment]
That would be a behavior change. The problem is that N was always
misleading. Perhaps we should adjust the documentation to say that it is N
"intervals", and therefore N+1 levels?

**Comment by jklymak** [comment]
Fair enough.  The PR #9472 is prob good enough in that case.

**Comment by tacaswell** [comment]
You may be surprised with how passionate peoples views on 'number of actual values' vs 'number of internals' is.....

**Comment by jklymak** [comment]
I was going to try not to be one of those people 😉

**Comment by jklymak** [comment]
Closed w/ #9472
